### PR TITLE
Do not process callbacks from cancelled LAVA jobs

### PR DESCRIPTION
**Pull request by mgalka**
LAVA keeps calling the callback API even tough the job is incomplete or
cancelled. Callbacks from canceled LAVA jobs may carry incomplete data
e.g. empty results section. This commit introduces changes that skip 
job processing when job status is different than 'complete' for test jobs 
and 'complete' or 'incomplete' for boot jobs.

### CODE REVIEW

File: app/handlers/callback.py [hunk 1]
```diff
@@ -68,13 +68,17 @@ def execute_put(self, *args, **kwargs):
     def _post(self, *args, **kwargs):
         try:
             lab_name = self.get_query_argument(models.LAB_NAME_KEY)
+            status_string = self.get_argument(models.LAVA_STATUS_STR_KEY)
```
Comment (gctucker) [inline app/handlers/callback.py]: This is LAVA-specific so it shouldn't be done in the main `CallbackHandler` class.  It should instead be done in `LavaCallbackHandler` or in the task handling the callback.

File: app/handlers/callback.py [hunk 2]
```diff
@@ -68,13 +68,17 @@ def execute_put(self, *args, **kwargs):
...
             req_token = kwargs["token"]
             valid_lab, error = self._is_valid_token(req_token, lab_name)
             if not valid_lab:
                 response = hresponse.HandlerResponse(403)
                 response.reason = (
                     "Provided authentication token is not associated with "
                     "lab '%s' or is not valid" % lab_name)
+            elif status_string == models.LAVA_CANCELED_STATUS_STR:
+                response = hresponse.HandlerResponse(400)
```
Comment (alexandrasp) [inline app/handlers/callback.py]: @mgalka is there any case that you will check is the status is incomplete? Just asking, cause you created a static  variable(LAVA_INCOMPLETE_STATUS_STR) but seems that you don't use. So, when you have a incomplete status, can you carry incomplete data? does it makes any sense?

File: app/utils/callback/lava.py [hunk 3]
```diff
@@ -645,7 +645,7 @@ def add_tests(job_data, job_meta, lab_name, db_options,
             ret_code, plan_doc_id, err = \
                 utils.kci_test.import_and_save_kci_tests(plan, db_options)
             utils.errors.update_errors(errors, err)
-    except (yaml.YAMLError, ValueError) as ex:
+    except (yaml.YAMLError, ValueError, KeyError) as ex:
```
Comment (gctucker) [inline app/utils/callback/lava.py]: This seems to be for different kind of fix, i.e. when anything is missing from the dictionary rather than checking the job status.  It should be in a separate commit.

File: app/handlers/callback.py (hunk 2 above)
Comment (mgalka) [reply app/handlers/callback.py]: It was added for the sake of consistency. Although it wasn't used it was put there to reflect all possible statues. Nevertheless, I'm handling the issue differently at the moment.

File: app/utils/callback/lava.py (hunk 3 above)
Comment (mgalka) [reply app/utils/callback/lava.py]: ok, removed from the current PR

File: app/handlers/callback.py (hunk 1 above)
Comment (mgalka) [reply app/handlers/callback.py]: moved handling code to lava.py

File: app/utils/callback/lava.py [hunk 4]
```diff
@@ -601,7 +608,9 @@ def add_tests(job_data, job_meta, lab_name, db_options,
     }
 
     try:
-        _get_job_meta(meta, job_data)
+        if not _get_job_meta(meta, job_data):
+            ret_code = 400
+            msg = "Unsupported job metadata"
```
Comment (gctucker) [inline app/utils/callback/lava.py]: But then, the rest still gets run even though we know we want to discard this job, right?

Comment (mgalka) [reply app/utils/callback/lava.py]: Right, I was looking for a solution that wouldn't require too many changes in the current error handling solution. The other idea was raising a `ValueError` in the `_get_job_meta()`, which is not quite what should be done here, too.
None of these solutions seem 100% good in this situation.

**Comment by mgalka** [comment]
Recent changes tested locally
```
[    INFO/MainThread] Received task: lava-test[ce1d7851-0962-4c56-abe4-f89c09d3191f]
[ WARNING/MainThread] Skipping LAVA job due to unsupported status: canceled
[    INFO/MainThread] Received task: test-regressions[479c1e8b-d00b-4fb1-9385-1cdc4bcb5277]
[ WARNING/MainThread] Not searching regressions as no test group ID
[    INFO/MainThread] Task test-regressions[479c1e8b-d00b-4fb1-9385-1cdc4bcb5277] succeeded in 0.00265978499965s: (200, None)
[    INFO/MainThread] Task lava-test[ce1d7851-0962-4c56-abe4-f89c09d3191f] succeeded in 0.0215948689997s: None
[    INFO/MainThread] Received task: lava-test[2d061f26-26e3-44c0-9539-cd7deb6374ea]
...
[    INFO/MainThread] Received task: lava-boot[05edf179-0292-449b-a71e-83c1567af537]
[ WARNING/MainThread] Skipping LAVA job due to unsupported status: canceled
[    INFO/MainThread] Received task: boot-regressions[c369fa39-9dea-40ac-9329-32fe06243c30]
[    INFO/MainThread] Searching boot regressions for 'None'
[    INFO/MainThread] Task lava-boot[05edf179-0292-449b-a71e-83c1567af537] succeeded in 0.027215954s: None
[    INFO/MainThread] No boot doc
[    INFO/MainThread] Task boot-regressions[c369fa39-9dea-40ac-9329-32fe06243c30] succeeded in 0.00827645799973s: None
...
[    INFO/MainThread] Received task: lava-test[96624d35-2418-40a9-84a9-32612835e81e]
[ WARNING/MainThread] Skipping LAVA job due to unsupported status: incomplete
[    INFO/MainThread] Received task: test-regressions[8e4d12ed-ba5f-4453-bbba-cf9e82dd3a90]
[ WARNING/MainThread] Not searching regressions as no test group ID
[    INFO/MainThread] Task test-regressions[8e4d12ed-ba5f-4453-bbba-cf9e82dd3a90] succeeded in 0.00169292500004s: (200, None)
[    INFO/MainThread] Task lava-test[96624d35-2418-40a9-84a9-32612835e81e] succeeded in 0.023454355s: None
```

File: app/utils/callback/lava.py [hunk 5]
```diff
@@ -370,6 +370,11 @@ def add_boot(job_data, job_meta, lab_name, db_options,
     :type base_path: string
     :return ObjectId The boot document id.
     """
+    if job_data.get("status") not in (COMPLETE, INCOMPLETE):
```
Comment (gctucker) [inline app/utils/callback/lava.py]: It would have been a bit better to add this test after the "Processing LAVA boot data..." message, to know which job is being skipped in the log.

File: app/utils/callback/lava.py [hunk 6]
```diff
@@ -584,6 +589,11 @@ def add_tests(job_data, job_meta, lab_name, db_options,
     :type base_path: string
     :return The top-level test group document id as ObjectId object.
     """
+    if job_data.get("status") not in (COMPLETE,):
```
Comment (gctucker) [inline app/utils/callback/lava.py]: Same thing here about log messages.

Also, I think we may actually not be able to do the `login` test case if we skip `INCOMPLETE` jobs, as if the login failed I believe the LAVA job status is going to be `INCOMPLETE`.  It's something worth verifying though.

**Comment by gctucker** [comment]
Tested OK on staging with a job I manually cancelled:
```
Nov 29 14:10:03 ip-172-30-0-172 kernelci-celery[31396]: [    INFO/MainThread] Processing LAVA test data: job 2076093 from lab-collabora
Nov 29 14:10:03 ip-172-30-0-172 kernelci-celery[31396]: [ WARNING/MainThread] Skipping LAVA job due to unsupported status: canceled
```
